krabs/krabs.cpp

```diff
@@ -4,12 +4,13 @@
 #include "krabs.hpp"
 
 // Note: This file exists to define the schema_locator
-// cache static variable. Due to issues with C++14 magic
+// cache static variables. Due to issues with C++14 magic
 // static and C++/CLI, we're unable to rely solely on
 // using magic static.
 
 // DO NOT ADD THINGS TO THIS FILE
 
 namespace krabs {
-    schema_locator schema_locator::singleton_;
-}
\ No newline at end of file
+    DWORD schema_locator::index_ = 0;
```

Comment:
@kallanreed's C++ is most definitely better than mine these days, but this seems like a reasonable opportunity to use the `thread_local` storage specifier? I think it'd also allow you to get around having to explicitly use the TlsAlloc/TlsSetValue functions, giving the code a more natural feel.

I defer to @kallanreed in case there are gotchas or bears here. It just seemed like a natural place to use it.

More documentation: https://en.cppreference.com/w/cpp/language/storage_duration

Reply:
Using the built-in thread local would be preferred so that you wouldn’t need to manage the locator cache memory at all - provided that works properly with managed c++. Off the top of my head, I don’t remember if this header is part of the “managed” part of the assembly in HIDS.

One thing I just remembered and you’ll definitely want to benchmark: this may add a pinvoke call from managed runtime to native for TlsGetData. We definitely need to be careful about that because the thunk can be expensive. IIRC that’s part of the reason we had added the cache in the first place. When krabs is compiled into managed C++ the call to tdh_get_schema was expensive because it was pinvoked.

Some additional docs here: https://docs.microsoft.com/en-us/cpp/parallel/thread-local-storage-tls?view=vs-2019

https://docs.microsoft.com/en-us/cpp/cpp/storage-classes-cpp?view=vs-2019#thread_local

Reply:
I'll have another look at thread_local - but my initial attempt bumped into
> error C2483: 'krabs::schema_locator::singleton_': object with destructor or non-aggregate type cannot be declared with thread storage duration in managed code

re: performance
We can probably avoid the pinvoke in the single-threaded single trace case if that helps?
The alternative approach would be to create a per-trace schema_locator and inject it appropriately - but the comments indicate that this was non-trivial to achieve.